LEV profile for display and example added

Sepp62 opened the pull request:
Hi Curtis, tried again with LEV profile from my fork

Sepp62 committed:
LEV profile for display and example added

cujomalainey commented on .gitignore:
```diff
@@ -8,3 +8,15 @@ src/main.cpp
 .DS_Store
 .tags*
 tags
+examples/LEVDisplay/__vm/.LEVDisplay.vsarduino.h
```
If we are going to merge the gitignore it would be great if we could generalize these files, such as 

`*.vsarduino.*` and `*.vmps.xml`

Sepp62 replied:
It's better not to merge gitignore...

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,217 @@
+	lev.onLevSpeedDist1(levSpeedDist1Handler);
+	lev.onLevSpeedDist2(levSpeedDist2Handler);
+	lev.onLevSpeedSysinfo2(levSpeedSysinfo2Handler);
+	lev.onLevBatteryInfo(levBatteryInfo);
```
Indentation is a bit messed up here

Sepp62 replied:
Done

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,217 @@
+    Serial.print("DataPage: ");
+    Serial.println(dp.getDataPageNumber());
+
+	// debug
```
Mind removing this section if its debug?

Sepp62 replied:
removed

cujomalainey commented on src/Profiles/LEV/Display/ANTPLUS_ProfileLevDisplay.h:
```diff
@@ -0,0 +1,53 @@
+	void onLevSpeedDist2(void(*func)(LevSpeedDist2&, uintptr_t), uintptr_t data = 0) { _onLevSpeedDist2.set(func, data); }
+	void onLevSpeedSysinfo2(void(*func)(LevSpeedSysinfo2&, uintptr_t), uintptr_t data = 0) { _onLevSpeedSysinfo2.set(func, data); }
+	void onLevBatteryInfo(void(*func)(LevBatteryInfo&, uintptr_t), uintptr_t data = 0) { _onLevBatteryInfo.set(func, data); }
+	void onLevCaps(void(*func)(LevCaps&, uintptr_t), uintptr_t data = 0) { _onLevCaps.set(func, data); }
```
Any chance can we use the full name here so it matches the data sheet and prevents confusion with the readers?

cujomalainey replied:
Also indentation

Sepp62 replied:
Names refactored, Indentation unclear (changed my editor from tabs to spaces)

cujomalainey replied:
Ah, its probably stuck on tabs, whitespace is awful to fix, I can send a pull request onto your branch with fixes, usually sublime text is the easiest way to convert it since it has built in utilities.

Sepp62 replied:
Done and checked. Visual Studio editor adjusted to spaces instead tabs

cujomalainey commented on src/Profiles/LEV/ANTPLUS_LevPrivateDefines.h:
```diff
@@ -0,0 +1,25 @@
+
+
+// Base page */
+#define ANTPLUS_LEV_DATAPAGEBASE_DATAPAGE_MASK 0x7F
```
Are you sure about this? What version of the datasheet are you using? In mine it uses the full byte for the datapage.

Sepp62 replied:
Removed

cujomalainey commented on src/Profiles/LEV/DataPages/Base/ANTPLUS_LevBaseMainDataPage.h:
```diff
@@ -0,0 +1,14 @@
+
+class LevBaseMainDataPage : public BaseDataPage<BroadcastData> {
+public:
+	LevBaseMainDataPage(AntRxDataResponse& dp);
```
Indentation

Sepp62 replied:
Done

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,217 @@
+
+void levBatteryInfo(LevBatteryInfo& msg, uintptr_t data) {
+	Serial.print("Charging Cycle Count: ");
+	Serial.println(msg.getChargingCycleCount());
```
Can we add a case for the unknowns special value? If not then lets add a TODO just so its not forgotten.

Sepp62 replied:
TODO added

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevBatteryInfo.cpp:
```diff
@@ -0,0 +1,27 @@
+
+uint16_t LevBatteryInfo::getChargingCycleCount()
+{
+	return (uint16_t)getData(2) + (((uint16_t)getData(3) & 0x0F) << 8);
```
This will need to be moved preprocessor defines eventually, but that can be done later by me, this is fine for now.

Sepp62 replied:
Left it as it was. Please add defines later.

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevBatteryInfo.cpp:
```diff
@@ -0,0 +1,27 @@
+
+uint16_t LevBatteryInfo::getFuelConsumption()
+{
+	uint16_t c = (uint16_t)getData(3) + (((uint16_t)getData(4) & 0x0F) << 8);
```
I think there is a bug here, I think it should be ((getData(3) & 0xF0) << 4) + getData(4)

Sepp62 replied:
Bug fixed, but not yet tested

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevBatteryInfo.cpp:
```diff
@@ -0,0 +1,27 @@
+
+uint16_t LevBatteryInfo::getDistanceOnCurrentCharge()
+{
+	return (uint16_t)getData(6) + (((uint16_t)getData(7) & 0x0F) << 8);
```
I don't see in the datasheet where it specifies its only half of byte 7, are you sure this is right?

Sepp62 replied:
Corrected

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,217 @@
+
+void levCaps(LevCaps& msg, uintptr_t data) {
+	Serial.print("Travel modes supported: ");
+	Serial.println(msg.getTravelModesSupported());
```
TODO for the bitfield

Sepp62 replied:
TODO added

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedDist.h:
```diff
@@ -0,0 +1,23 @@
+class LevSpeedDist1 : public LevBaseMainDataPage {
+public:
+    LevSpeedDist1(AntRxDataResponse& dp);
+	uint32_t getTotalDist(); // in km
```
Indentation

Sepp62 replied:
Done

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedDist.cpp:
```diff
@@ -0,0 +1,42 @@
+	return;
+}
+
+uint32_t LevSpeedDist1::getTotalDist() // in km
```
Can we call this Odometer? Just so it matches the datasheet and its easier to look up?

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedDist.cpp:
```diff
@@ -0,0 +1,42 @@
+	return;
+}
+
+uint32_t LevSpeedDist2::getTotalDist() // in km
```
Same request as above.

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,217 @@
+
+void levSpeedSysinfo1Handler(LevSpeedSysinfo1& msg, uintptr_t data) {
+	Serial.print("Temp state: ");
+	Serial.println(msg.getTempState());
```
TODO for state enum

Sepp62 replied:
TODO added

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedSysinfo1.cpp:
```diff
@@ -0,0 +1,36 @@
+	return;
+}
+
+uint8_t LevSpeedSysinfo1::getTempState()
```
Lets use TemperatureState so its not confused with "TemporaryState"

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedSysinfo1.cpp:
```diff
@@ -0,0 +1,36 @@
+	return getData(4);
+}
+
+uint8_t LevSpeedSysinfo1::getError()
```
Lets use getErrorMessage, might get confused with getError in the framework.

Sepp62 replied:
Done

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedSysinfo2.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <Profiles/Lev/DataPages/RX/ANTPLUS_LevSpeedSysinfo2.h>
+#include <Profiles/Lev/ANTPLUS_LevPrivateDefines.h>
+
+LevSpeedSysinfo2::LevSpeedSysinfo2(AntRxDataResponse& dp) : LevBaseMainDataPage(dp) {
```
Camel case error Sys**I**nfo

Sepp62 replied:
Names refactored

cujomalainey commented on src/Profiles/LEV/DataPages/RX/ANTPLUS_LevSpeedSysinfo2.h:
```diff
@@ -0,0 +1,17 @@
+
+#include <Profiles/Lev/DataPages/Base/ANTPLUS_LevBaseMainDataPage.h>
+
+class LevSpeedSysinfo2 : public LevBaseMainDataPage {
```
I wonder if its worth making a special intermediate class for SpeedSysInfo and SpeedDistanceInfo so all the fields are not duplicated.

Sepp62 replied:
Left it unchanged. Please change it later, if you like.

cujomalainey commented on src/Profiles/LEV/Display/ANTPLUS_ProfileLevDisplay.cpp:
```diff
@@ -0,0 +1,120 @@
+    uint8_t dataPage = dp.getDataPageNumber();
+    bool called = false;
+
+	switch (dataPage) {
```
Indentation

Sepp62 replied:
Done

cujomalainey commented on src/Profiles/LEV/Display/ANTPLUS_ProfileLevDisplay.cpp:
```diff
@@ -0,0 +1,120 @@
+		called = handleCaps(dp);
+		break;
+	case ANTPLUS_COMMON_DATAPAGE_MANUFACTURERSINFORMATION_NUMBER:
+		// Called Manufacturers Identification in Environment Profile PDF
```
Not needed here, but appreciated :)

Sepp62 replied:
Unnecessary comment removed

cujomalainey reviewed (changes requested):
Great first run, mostly minor stuff.

cujomalainey commented:
Oh also can you add the example to the travis config file so it gets build and tested in the PR?

Sepp62 committed:
Refactoring Part 1

cujomalainey committed:
Convert all tabs to spaces

Repo uses 4 space indentation, even in 2019 indentation conversion is
still hard :P thank you sublime text for saving us.

cujomalainey commented on src/Profiles/ANTPLUS_Profiles.h:
```diff
@@ -4,5 +4,6 @@
 #include <Profiles/BicycleSpeed/ANTPLUS_BicycleSpeedProfile.h>
 #include <Profiles/HeartRate/ANTPLUS_HeartRateProfile.h>
 #include <Profiles/Environment/ANTPLUS_EnvironmentProfile.h>
+#include <Profiles/Lev/ANTPLUS_LevProfile.h>
```
Looks like you have some capitalization differences here, looks like travis caught them
https://travis-ci.org/cujomalainey/antplus-arduino/jobs/544454378
(another joy is OSX and windows are case insensitive about paths but linux isn't)

Sepp62 replied:
Done

cujomalainey committed:
LEV: Display: Add example to Travis config

Add to build so its compile tested all the time

Sepp62 committed:
Merge pull request #1 from cujomalainey/temp/remote/pr

Convert all tabs to spaces

Sepp62 committed:
rename LEV folder to Lev

Sepp62 committed:
getError() --> getErrorMessage()

Sepp62 committed:
Unnecessary comment removed

Sepp62 committed:
Some corrections, mainly TODOs added and some decoding bug fixes

Decoding not yet completly tested (TODOS added)

Sepp62 committed:
.gitignore rolled back

Sepp62 commented:
All issues handled best to my knowledge. Please feel free to adapt the code as you like.

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,207 @@
+
+#define BAUD_RATE 9600
+#define CHANNEL_0 0
+#define CHANNEL_1 1
```
this isnt used anywhere, best to remove it so we don't confuse anyone

Sepp62 replied:
Removed

cujomalainey commented on examples/LEVDisplay/LEVDisplay.ino:
```diff
@@ -0,0 +1,207 @@
+    Serial.print("Gear error: ");
+    Serial.println(msg.getErrorMessage());          // TODO enums for error message
+    Serial.print("Speed: ");
+    Serial.print(msg.getSpeed()/10);
```
if I am correct, all these getSpeed calls are getting truncated due to integer division, / 10.0 should fix it, make sure to check the rest of the divisions :)

Sepp62 replied:
There are two print statements, the first one for integer and the second one for the fraction. Of course you can use float calculation also, but this would be the only place in the whole program which uses floating point data types. I dont know exactly, what all the arduino toolchains will do, but they probably will link the floating point runtime library, due to this statement.

Feel free to change it in your branch.

cujomalainey replied:
Yep you are right, i totally missed that.

cujomalainey commented on src/Profiles/Lev/ANTPLUS_LevPrivateDefines.h:
```diff
@@ -0,0 +1,24 @@
+#define ANTPLUS_LEVPROFILEPRIVATEDEFINES_h
+
+/* Channel Config */
+#define ANTPLUS_LEV_CHANNELTYPE CHANNEL_TYPE_BIDIRECTIONAL_RECEIVE
```
Lets make this ANTPLUS_LEV_DISPLAY_CHANNELTYPE so we know which side of the profile we are talking about

Sepp62 replied:
Changed

cujomalainey commented on src/Profiles/Lev/Display/ANTPLUS_ProfileLevDisplay.cpp:
```diff
@@ -0,0 +1,119 @@
+}
+
+/*
+void ProfileLevDisplay::onChannelEventResponse(ChannelEventResponse& msg) {
```
More debug?

Sepp62 replied:
Removed. I made this just to show a potential user how to receive channel event notifications.

cujomalainey replied:
Ah i would make an example for that, I wouldn't expect anyone to dive into the backend to find that

cujomalainey commented on src/Profiles/Lev/Display/ANTPLUS_ProfileLevDisplay.h:
```diff
@@ -0,0 +1,53 @@
+    void onProductInformation(void(*func)(ProductInformation&, uintptr_t), uintptr_t data = 0) { _onProductInformation.set(func, data); }
+
+protected:
+    // virtual void onChannelEventResponse(ChannelEventResponse& msg);
```
more debug?

Sepp62 replied:
Removed

cujomalainey reviewed (changes requested):
Looking really good, just a couple more comments then I'll merge this. I really appreciate the detail you took to naming.

cujomalainey commented:
Also, make sure to leave the pull request open, Travis CI can't pull the code to test this if it gets closed, closing is like a "cancel request" marker if it hasn't been merged yet and I definitely want to see this go in

Sepp62 committed:
Constant changed and commented code removed

cujomalainey reviewed (approved):
Looking good
